python_modules/dagster/dagster/_core/definitions/asset_data_sla.py

```diff
@@ -0,0 +1,113 @@
+from typing import Mapping, Optional
+
+import pendulum
+from croniter import croniter
+
+from .events import AssetKey
+
+
+class RootDataTimeConstraint:
+    """Specifies a constraint on the timestamp of the root data that was used for the most recent
+    asset materialization of a given asset.
+
+    The root data timestamp is calculated by traversing the history of asset materializations of
+    upstream assets which occured before the current materialization. This gives a lower bound on
+    the most recent records that could possibly be incorporated into the current state of the asset
+    to which this SLA is attached.
+    """
+
+    def is_passing(self, root_data_timestamp: Optional[float]) -> bool:
+        pass
+
+
+class StalenessConstraint(RootDataTimeConstraint):
+    """This constraint specifies that the root data that was used for the most recent asset
+    materialization must not be more than `allowed_staleness_minutes` old, relative to the current
+    time.
+    """
+
+    def __init__(self, allowed_staleness_minutes: float):
+        self._allowed_staleness = pendulum.duration(minutes=allowed_staleness_minutes)
+
+    def is_passing(self, root_data_timestamp: Optional[float]) -> bool:
+        if root_data_timestamp is None:
+            return False
+
+        return (
+            pendulum.from_timestamp(root_data_timestamp) > pendulum.now() - self._allowed_staleness
+        )
+
+
+class CronStalenessConstraint(RootDataTimeConstraint):
+    """This constraint specifies that the root data that was used for the most recent asset
+    materialization must not be more than `allowed_staleness_minutes` old, relative to the most
+    recent cron schedule tick.
+    """
+
+    def __init__(self, cron_schedule: str, allowed_staleness_minutes: float):
+        self._cron_schedule = cron_schedule
+        self._allowed_staleness = pendulum.duration(minutes=allowed_staleness_minutes)
+
+    def is_passing(self, root_data_timestamp: Optional[float]) -> bool:
```

Comment:
Thoughts on floats vs. pendulum timestamps?

Reply:
decided to keep things as floats, just to not bake pendulum too deeply into the APIs here. Theoretically, you can use whatever time library you want to calculate these things (although I don't see a reason that would ever be anything other than pendulum).

don't feel strongly about this and can be swayed otherwise